inference/src/translator/text_processor.cpp

```diff
@@ -106,6 +106,117 @@ void TextProcessor::process(std::string &&input, AnnotatedText &source, Segments
     }
   }
 }
+#elif defined(WASM)
+void TextProcessor::process(std::string&& input, AnnotatedText& source, Segments& segments) const {
+  /**
+   * This is an RAII guard for auto-freeing a malloc-allocated pointer when it leaves scope.
+   *
+   * The address of the internal pointer is passed to JavaScript where it is allocated
+   * via `malloc`, however we are responsible for deallocating the memory via `free` on
+   * the C++ side.
+   *
+   * Wrapping the deallocation in the checked destructor ensures that once the memory has been 
+   * allocated on the JS side, it is guaranteed to be deallocated when this stack frame is popped,
+   * whether from returning or from an exception being thrown.
+   */
+  struct ScopedPtr {
+    int32_t* ptr = nullptr;
+
+    const int32_t operator[](const size_t index) const {
+      return ptr[index];
+    }
+
+    ~ScopedPtr() {
+      if (ptr) {
+        free(ptr);
+        ptr = nullptr;
+      }
+    }
+  };
+
+  ScopedPtr starts;
+  ScopedPtr ends;
+  int32_t sentenceCount = 0;
+
+  source = std::move(AnnotatedText(std::move(input)));
+  std::string_view input_converted(source.text.data(), source.text.size());
+
+  EM_ASM({
+    // Attach a lazily initialized cache of the Intl.Segmenter to the WASM Module.
+    // This is critical to prevent creating one instance of Intl.Segmenter per
+    // translation request.
+    if (!Module.getOrCreateSentenceSegmenter) {
+      Module.getOrCreateSentenceSegmenter = (function() {
+        let segmenters = new Map();
+
+        return function(lang) {
+          if (!segmenters.has(lang)) {
```

Comment:
Nitpick: This may not matter, but since this is performance critical code, this could remove a lookup, and is the general pattern I use for this kind of code.

```js
let segmenter = segmenters.get(lang);
if (!segmenter) {
  segmenter = new Intl.Segmenter(lang, { granularity: "sentence" });
  segmenters.set(lang, segmenter);
}
return segmenter;
```